File: android/app/src/main/java/com/example/demo/data/DemoRepository.kt
package com.example.demo.data

import dev.openfeature.kotlin.sdk.Client
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class DemoRepository
    @Inject
    constructor(
        private val client: Client,
    ) {
        fun getWelcomeMessage(): Flow<String> =
            flow {
                emit("Choose a plan")
            }

        fun getPlanConfiguration(): Flow<PlanConfiguration> =
            flow {
                val enterpriseConfigEnabled = client.getBooleanValue("show-enterprise-plan.enabled", false)
                val enterpriseConfigPrice = client.getStringValue("show-enterprise-plan.price", "Contact Us")
                val highlightPremium =
                    client.getBooleanValue(
                        "subscription-highlight.enabled",
                        false,
                    )

                emit(
                    PlanConfiguration(
                        showEnterprisePlan = enterpriseConfigEnabled,
                        enterprisePrice = enterpriseConfigPrice,
                        highlightPremium = highlightPremium,
                    ),
                )
            }

        fun getSubscriptionPlans(config: PlanConfiguration): Flow<List<SubscriptionPlan>> =
            flow {
                val basePlans =
                    listOf(
                        SubscriptionPlan(
                            name = "Basic",
                            price = "$9",
                            features =
                                listOf(
                                    "Access to Feature A",
                                    "Standard support",
                                    "Basic usage limits",
                                    "Email notifications",
                                ),
                        ),
                        SubscriptionPlan(
                            name = "Premium",
                            price = "$19",
                            features =
                                listOf(
                                    "All Basic features",
                                    "Access to Feature B",
                                    "Priority support",
                                    "Advanced usage limits",
                                    "API access",
                                ),
                            isHighlighted = config.highlightPremium,
                        ),
                        SubscriptionPlan(
                            name = "Pro",
                            price = "$39",
                            features =
                                listOf(
                                    "All Premium features",
                                    "Access to Feature C",
                                    "Premium support",
                                    "Highest usage limits",
                                    "Custom integrations",
                                    "Advanced analytics",
                                ),
                        ),
                    )

                val plans =
                    if (config.showEnterprisePlan) {
                        basePlans +
                            SubscriptionPlan(
                                name = "Enterprise",
                                price = config.enterprisePrice,
                                features =
                                    listOf(
                                        "All Premium features",
                                        "Unlimited usage",
                                        "Dedicated account manager",
                                        "Custom solutions",
                                        "SLA guarantee",
                                    ),
                            )
                    } else {
                        basePlans
                    }

                emit(plans)
            }
    }

File: ios/ConfidenceDemoApp/Data/DemoRepository.swift
import Foundation
import Combine
import OpenFeature

class DemoRepository: ObservableObject {
    private let openFeatureClient = OpenFeatureAPI.shared.getClient()

    func getWelcomeMessage() -> AnyPublisher<String, Never> {
        return Future { promise in
            Task {
                // This flag doesn't exist so we expect to always fall back to defaults
                let welcomeMessage = self.openFeatureClient.getStringValue(key: "welcome-message", defaultValue: "Welcome to Confidence Demo")
                promise(.success(welcomeMessage))
            }
        }
        .eraseToAnyPublisher()
    }

    func getPlanConfiguration() -> AnyPublisher<PlanConfiguration, Never> {
        return Future { promise in
            Task {
                let showEnterprise = self.openFeatureClient.getBooleanValue(key: "show-enterprise-plan", defaultValue: false)
                let enterprisePrice = self.openFeatureClient.getStringValue(key: "enterprise-price", defaultValue: "Contact Sales")
                let highlightPremium = self.openFeatureClient.getBooleanValue(key: "highlight-premium", defaultValue: false)

                let config = PlanConfiguration(
                    showEnterprisePlan: showEnterprise,
                    enterprisePrice: enterprisePrice,
                    highlightPremium: highlightPremium
                )
                promise(.success(config))
            }
        }
        .eraseToAnyPublisher()
    }

    func getSubscriptionPlans(config: PlanConfiguration) -> AnyPublisher<[SubscriptionPlan], Never> {
        return Future { promise in
            var plans: [SubscriptionPlan] = []

            // Basic Plan
            plans.append(SubscriptionPlan(
                name: "Basic",
                price: "$9.99/month",
                features: [
                    "Access to basic features",
                    "Email support",
                    "1 user account"
                ]
            ))

            // Premium Plan
            plans.append(SubscriptionPlan(
                name: "Premium",
                price: "$19.99/month",
                features: [
                    "All basic features",
                    "Priority support",
                    "5 user accounts",
                    "Advanced analytics"
                ],
                isHighlighted: config.highlightPremium
            ))

            // Enterprise Plan (conditional)
            if config.showEnterprisePlan {
                plans.append(SubscriptionPlan(
                    name: "Enterprise",
                    price: config.enterprisePrice,
                    features: [
                        "All premium features",
                        "24/7 dedicated support",
                        "Unlimited users",
                        "Custom integrations",
                        "SLA guarantee"
                    ]
                ))
            }

            promise(.success(plans))
        }
        .eraseToAnyPublisher()
    }
}
